core/lib/spree/event/subscriber.rb

```diff
@@ -54,32 +54,35 @@ def self.included(base)
       #     end
       #   end
       def event_action(method_name, event_name: nil)
-        mattr_accessor "#{method_name}_handler"
+        mattr_writer "#{method_name}_handler"
+
+        define_method "#{method_name}_handler" do
+          Spree::Deprecation.warn("#{name}.#{method_name}_handler and #{name}.#{method_name}_handler= from the old events mapping interface are deprecated. Please use the new mapping stored in Spree::Event.subscribers.", caller)
+
+          class_variable_get("@@#{method_name}_handler")
+        end
+
         event_actions[method_name] = (event_name || method_name).to_s
       end
 
-      # Subscribes all declared event actions to their events. Only actions that are subscribed
+      # @deprecated Subscribes all declared event actions to their events. Only actions that are subscribed
       # will be called when their event fires.
       #
       # @example subscribe all event actions for module 'EmailSender'
       #    EmailSender.subscribe!
       def subscribe!
         unsubscribe!
-        event_actions.each do |event_action, event_name|
-          send "#{event_action}_handler=", Spree::Event.subscribe(event_name) { |event|
-            send event_action, event
-          }
-        end
+        Spree::Deprecation.warn("#{name}.subscribe! is deprecated. Please move your event subscribers to folder `app/subscribers/*_subscriber.rb` and they will be loaded automatically. ", caller)
+        Spree::Event.subscribers.send(:subscribe, self)
       end
 
-      # Unsubscribes all declared event actions from their events. This means that when an event
+      # @deprecated Unsubscribes all declared event actions from their events. This means that when an event
       # fires then none of its unsubscribed event actions will be called.
       # @example unsubscribe all event actions for module 'EmailSender'
       #    EmailSender.unsubscribe!
       def unsubscribe!
-        event_actions.keys.each do |event_action|
-          Spree::Event.unsubscribe send("#{event_action}_handler")
-        end
+        Spree::Deprecation.warn("#{name}.unsubscribe! is deprecated. Please move your event subscribers to folder `app/subscribers/*_subscriber.rb` and they will be unloaded automatically. ", caller)
```

Comment:
I think the message here is incorrect. You have to delete the subscriber file to unsubscribe, isn't it? Also, I'd keep the method since maybe you don't have control over where the file has been added (core or an extension).

Reply:
Yes, you're right about the message. And also the remainder of the comment is right. I think we need to keep offering that kind of granularity, even though it will be rarely used.